### PR TITLE
When displaying multispans, ignore empty lines adjacent to `...`

### PR DESCRIPTION
**Pull request by estebank**
```
error[E0308]: `match` arms have incompatible types
   --> tests/ui/codemap_tests/huge_multispan_highlight.rs:98:18
    |
6   |       let _ = match true {
    |               ---------- `match` arms have incompatible types
7   |           true => (
    |  _________________-
8   | |             // last line shown in multispan header
...   |
96  | |
97  | |         ),
    | |_________- this is found to be of type `()`
98  |           false => "
    |  __________________^
...   |
119 | |
120 | |         ",
    | |_________^ expected `()`, found `&str`

error[E0308]: `match` arms have incompatible types
   --> tests/ui/codemap_tests/huge_multispan_highlight.rs:215:18
    |
122 |       let _ = match true {
    |               ---------- `match` arms have incompatible types
123 |           true => (
    |  _________________-
124 | |
125 | |         1 // last line shown in multispan header
...   |
213 | |
214 | |         ),
    | |_________- this is found to be of type `{integer}`
215 |           false => "
    |  __________________^
216 | |
217 | |
218 | |         1 last line shown in multispan
...   |
237 | |
238 | |         ",
    | |_________^ expected integer, found `&str`
```

### CODE REVIEW

File: tests/ui/codemap_tests/huge_multispan_highlight.svg
Comment (oli-obk) [inline tests/ui/codemap_tests/huge_multispan_highlight.svg:1]: 😆 nice to view, but please bless the test again and remove the svg

Comment (estebank) [reply tests/ui/codemap_tests/huge_multispan_highlight.svg:1]: I made it a `--color=always` test to ensure that the highlighting for multispans doesn't break. We don't have nearly enough tests for the color output as we stand (and I know we have some bugs on the path-diff suggestion rendering that I'll add to the test suite as soon as I manage to reproduce one).

Comment (estebank) [reply tests/ui/codemap_tests/huge_multispan_highlight.svg:1]: If you want I can turn this one to `.stderr` and introduce a new one for the colors?

Comment (oli-obk) [reply tests/ui/codemap_tests/huge_multispan_highlight.svg:1]: wait lol, --color=always makes ui tests produce svgs?!

Comment (estebank) [reply tests/ui/codemap_tests/huge_multispan_highlight.svg:1]: https://github.com/rust-lang/rust/pull/121877

Comment (oli-obk) [reply tests/ui/codemap_tests/huge_multispan_highlight.svg:1]: holy cow, and github even produces diffs

File: tests/rustdoc-ui/lints/check.stderr [hunk 1]
```diff
@@ -4,7 +4,6 @@ warning: missing documentation for the crate
 LL | / #![feature(rustdoc_missing_doc_code_examples)]
 LL | |
 LL | |
```
Comment (oli-obk) [inline tests/rustdoc-ui/lints/check.stderr:6]: why are there still empty lines above?

Comment (estebank) [reply tests/rustdoc-ui/lints/check.stderr:6]: They correspond to `//~ ERROR` annotations, which are removed from view, but that are not considered empty by the code.

File: compiler/rustc_errors/src/emitter.rs [hunk 2]
```diff
@@ -2435,7 +2474,16 @@ impl FileWithAnnotatedLines {
...
                 let middle = min(ann.line_start + 4, ann.line_end);
-                for line in ann.line_start + 1..middle {
+                // We'll show up to 4 lines past the beginning of the multispan start.
+                // We will *not* include the tail of lines that are only whitespace.
+                let until = (ann.line_start..middle)
+                    .rev()
+                    .filter_map(|line| file.get_line(line - 1).map(|s| (line + 1, s)))
+                    .filter(|(_, s)| !s.trim().is_empty())
+                    .map(|(line, _)| line)
+                    .next()
```
Comment (oli-obk) [inline compiler/rustc_errors/src/emitter.rs:2484]:
```suggestion
                    .find(|(_, s)| !s.trim().is_empty())
                    .map(|(line, _)| line)
```

**Review by oli-obk — CHANGES_REQUESTED** [review]
r=me after merging the fixup commits into their corresponding main commits

File: tests/ui/codemap_tests/huge_multispan_highlight.svg
Comment (nikomatsakis) [reply tests/ui/codemap_tests/huge_multispan_highlight.svg:1]: oooooomg

**Comment by workingjubilee** [comment]
@bors r-

Seems to have busted rollup in https://github.com/rust-lang/rust/pull/122260#issuecomment-1987021472

**Comment by estebank** [comment]
Oooff... Seeing what the diff was concerns me about relying on the svg output going forward :-/

**Comment by estebank** [comment]
@epage could you bring some insight on what the reason might be for this divergence in the output?

```
---- [ui] tests/ui/codemap_tests/huge_multispan_highlight.rs stdout ----
diff of svg:

- <svg width="743px" height="848px" xmlns="http://www.w3.org/2000/svg">
+ <svg width="740px" height="848px" xmlns="http://www.w3.org/2000/svg">
2   <style>
3     .fg { fill: #AAAAAA }
4     .bg { background: #000000 }
```

I suspected maybe font size differences causing some calculation to diverge?

**Comment by epage** [comment]
Not sure what is going on with a difference in width of 3px as nothing works on that small of scale.

**Comment by estebank** [comment]
@epage that's a shame. For now I think that setting a min-width unblocks us, but we should be aware of that issue. For humans it doesn't matter, but it will get in the way of tooling doing diffs (like the cargo test suite, also surprised it didn't happen there either).

**Comment by epage** [comment]
Cargo is getting this functionality through `snapbox` which has a (hacky) solution to only pay attention to the `<text>` and not to the stuff at the top so it gives a more focused diff to the user.

**Comment by estebank** [comment]
@epage ah! I see. That makes sense, to make the differ smarter and ignore what humans would ignore. For now I'll go with the min-width approach to see if that's enough, so that we can still detect changes in the output that would cause our bounding box to change, but if it continues to give us trouble I'll modify the diff checker to ignore the sizing info (shouldn't be too hard to do, it can even be "ignore first line").

**Comment by nikic** [comment]
The test still fails for me locally:
```
---- [ui] tests/ui/codemap_tests/huge_multispan_highlight.rs stdout ----
diff of svg:

-	<svg width="750px" height="848px" xmlns="http://www.w3.org/2000/svg">
+	<svg width="751px" height="848px" xmlns="http://www.w3.org/2000/svg">
2	  <style>
3	    .fg { fill: #AAAAAA }
4	    .bg { background: #000000 }
```

**Comment by estebank** [comment]
@nikic I'll make an emergency PR to completely ignore the width then :-/

**Pull request merged**